anndata/core/alignedmapping.py

```diff
@@ -75,12 +71,10 @@ def _validate_value(self, val: V, key: str) -> None:
                     f"{self.axes} of parent. Value had shape {val.shape} while "
                     f"it should have had {right_shape}."
                 )
-        try:  # TODO: Handle objects with indices
-            # Could probably also re-order index if it's contained
-            if not (val.index == self.dim_names).all():
-                raise IndexError()  # Maybe not index error
-        except AttributeError:
-            pass
```

Comment:
@flying-sheep, why did this get removed?

Reply:
It wasn’t, ctrl+f “val.index == self.dim_names”:

https://github.com/theislab/anndata/pull/228/files#diff-2a3b9403444574f42d160b2f5ea71bc4R221

Reply:
But why not keep it here? Arguably the same `super()` dispatch could also happen with the homogenous df. In either case, I think the shape check error is more helpful when it applies, and should happen first.

Reply:
Agreed about the second part, but I didn’t understand the first: Where could it maybe happen?

Reply:
For the first part, the Layers and Pairwise classes could get a method like this:

```python
def _validate_value(self, val: V, key: Hashable) -> V:
    if not self._allow_df and isinstance(val, pd.DataFrame):
        name = self.attrname.title().rstrip('s')
        val = ensure_df_homogeneous(val, f'{name} {key!r}')
    return super()._validate_value(val, key)
```

Not that I think this is the way to go. This feels a little "aspect-y", which is not a common paradigm in our codebase.